spec/index.bs

```diff
@@ -642,36 +682,83 @@ interface GPUAdapter {
 
     Requests a [=device=] from the [=adapter=].
 
-    Returns [=a new promise=], |promise|.
+    Returns [=a new promise=], |promise|. Then, the following occurs asynchronously:
 
-      - If the following requirements are met, |promise| [=resolves=] to a new
-        {{GPUDevice}} object encapsulating a new [=device=], which MUST
-        have exactly the capabilities described by |descriptor|.
+      - If the user agent can fulfill the request and
+        the [$requestDevice Valid Usage|valid usage$] rules are met:
 
-        Issue: Figure out how to integrate the requirements.
+          - |promise| [=resolves=] to a new {{GPUDevice}} object encapsulating
+            [=a new device=] with the capabilities described by |descriptor|.
 
       - Otherwise, |promise| [=rejects=] with an {{OperationError}}.
 </div>
 
+<div algorithm class=validusage>
+    <dfn abstract-op>requestDevice Valid Usage</dfn>
+
+    Given an [=adapter=] |adapter| and a {{GPUDeviceDescriptor}} |descriptor|,
+    the following validation rules apply:
+
+      - The set of {{GPUExtensionName}} values in
+        |descriptor|.{{GPUDeviceDescriptor/extensions}}
+        must be a subset of those in |adapter|.{{adapter/[[extensions]]}}.
+
+      - For each type of limit in {{GPULimits}}, the value of that limit in
+        |descriptor|.{{GPUDeviceDescriptor/limits}}
+        must be no [=better=] than the value of that limit in
+        |adapter|.{{adapter/[[limits]]}}.
+</div>
+
+#### <dfn dictionary>GPUDeviceDescriptor</dfn> #### {#gpudevicedescriptor}
+
+{{GPUDeviceDescriptor}} describes a device request.
+
 <script type=idl>
 dictionary GPUDeviceDescriptor : GPUObjectDescriptorBase {
     sequence<GPUExtensionName> extensions = [];
     GPULimits limits = {};
-
-    // TODO: are other things configurable like queues?
 };
 </script>
 
+<dl dfn-type=dict-member dfn-for=GPUDeviceDescriptor>
+    : <dfn>extensions</dfn>
+    ::
+        The set of {{GPUExtensionName}} values in this sequence defines the exact set of
+        extensions that must be supported and allowed by a requested device.
+
+    : <dfn>limits</dfn>
+    ::
+        Defines the exact limits that must be supported and allowed by a requested device.
+</dl>
+
+#### <dfn enum>GPUExtensionName</dfn> #### {#gpuextensionname}
+
+Each {{GPUExtensionName}} identifies a set of functionality which, if available,
+allows additional usages of WebGPU that would have otherwise been invalid.
+
 <script type=idl>
 enum GPUExtensionName {
-    "anisotropic-filtering"
+    "texture-compression-bc"
 };
 </script>
 
+<dl dfn-type=enum-value dfn-for=GPUExtensionName>
+    : <dfn>"texture-compression-bc"</dfn>
+    ::
+        Issue: Write a spec section for this, and link to it.
+</dl>
+
+#### <dfn dictionary>GPULimits</dfn> #### {#gpulimits}
+
+{{GPULimits}} describes various limits in the usage of WebGPU on a device.
+
+One limit value may be <dfn dfn>better</dfn> than another.
+For each limit, "better" is defined.
```

Comment:
nice :+1: 
I am concerned though. In some cases, requesting higher limits could result in *slower* implementation, assuming the implementation needs to prepare space and logic to account for the limits...
So it's strange to call them "better" in this context :/

Reply:
Added a note about this, which I think is sufficient.